---
title: "Genomic Prediction"
output: html_notebook
---

&nbsp;


&nbsp;

#### Uploading R libraries

```{r}
library(BGLR)
library(ggplot2)
library(readr)
library(rcompanion)
library(readxl)
library(kableExtra)
library(GAPIT)
```

&nbsp;

#### Analysis Path

```{r}
setwd("D:/AGROSAVIA_ 2020 _BUHO/PAPERS_2020/Korea Case/Paper GP/A someter/American Naturalist/Scripts/Genomic Prediction")
```


&nbsp;

#### Uploading the Allelic Variant File 

```{r}
library(GAPIT)
# Using GAPIT for convert the hamap format to numeric format
# Set the folder relative to the variable in the path, in this case we use the folder "yield" as an example. Explore the folders: "NP" (number of pods), "NS" (average number of seeds per pod), "BS" (seed weight), and "VB" (vegetative biomass )
myG <- as.data.frame(read_delim("D:/AGROSAVIA_ 2020 _BUHO/PAPERS_2020/Korea Case/Paper GP/A someter/American Naturalist/Scripts/Genomic Prediction/Yield_KKNimp.hmp.txt",
                                delim = "\t", escape_double = FALSE,
                                col_names = FALSE, trim_ws = TRUE))
myGAPIT <- GAPIT(G=myG, output.numerical=TRUE,
                 Geno.View.output=FALSE)
# myG <- as.data.frame(read_delim("D:/AGROSAVIA_ 2020 _BUHO/PAPERS_2020/Korea Case/Paper GP/Genotype/Yield/Yield_GWAS_KKNimp.hmp.txt",
#                                 delim = "\t", escape_double = FALSE,
#                                 col_names = FALSE, trim_ws = TRUE))
myGAPIT <- GAPIT(G=myG, output.numerical=TRUE,
                 Geno.View.output=FALSE)

```

```{r}
# Set the folder relative to the variable in the path, in this case we use the folder "yield" as an example. Explore the folders: "NP" (number of pods), "NS" (average number of seeds per pod), "BS" (seed weight), and "VB" (vegetative biomass )
X <- read.delim("D:/AGROSAVIA_ 2020 _BUHO/PAPERS_2020/Korea Case/Paper GP/A someter/American Naturalist/Scripts/Genomic Prediction/GAPIT.Genotype.Numerical.txt", row.names=1)


X[X==2] <- -1
X[X==1] <- 7
X[X==0] <- 1
X[X==7] <- 0

X[1:10,1:10] %>%
  kbl(caption = "Allelic Variant ") %>%
  kable_classic(full_width = F, html_font = "Cambria")
```

&nbsp;

#### Uploading the Genomic Relationship matrix
```{r include=FALSE} 
G <- read_csv("D:/AGROSAVIA_ 2020 _BUHO/PAPERS_2020/Korea Case/Paper GP/A someter/American Naturalist/Scripts/Genomic Prediction/GAPIT.Kin.VanRaden.csv", 
              col_names = FALSE)
G <- as.matrix(G[,-1])
```

```{r}
G[1:5,1:5] %>%
  kbl(caption = "Genomic Relationship matrix") %>%
  kable_classic(full_width = F, html_font = "Cambria")
```

&nbsp;

Upload the phenotypic data related to variable "x", we use "Yield" as an example
```{r}
# Set the folder relative to the variable in the path, in this case we use the folder "yield" as an example. Explore the folders: "NP" (number of pods), "NS" (average number of seeds per pod), "BS" (seed weight), and "VB" (vegetative biomass )
YLP1 <- read_delim("D:/AGROSAVIA_ 2020 _BUHO/PAPERS_2020/Korea Case/Paper GP/A someter/American Naturalist/Scripts/Genomic Prediction/Index_YLP_Kolfaci.txt", 
    delim = "\t", escape_double = FALSE, 
    trim_ws = TRUE)
names <- as.data.frame(rownames(X))
colnames(names) <- c("Taxa")
YLP <- merge(YLP1, names, by="Taxa")
YLP <- YLP[,1:5]
YLP %>%
  kbl(caption = "Phenotypic Table Yield") %>%
  kable_classic(full_width = F, html_font = "Cambria")
```


&nbsp;

#### Cross-validation

&nbsp;

```{r}
# For example, Environment 2 refers to the second column related to the environment 'Motilonia'.
Environment <- 2
# 1, 2 or 3
y<-YLP[,Environment]
```

```{r}
set.seed(0000) # We use several seeds in the study to measure the stochastic outputs
sets <- sample(c(1,2,3,4,5), size = 83, replace = TRUE)
#Case III - Cross-validation 
for(i in 1:5){
  yNa=y
  whichNa=(sets==i)
  yNa[whichNa]=NA
  
  ETA=list(list(X=X,model="BayesC"))
  # Z2 <- as.matrix(X)
  # G2= tcrossprod(Z2)/ncol(Z2)
  # ETA=list(list(K=G2,model="RKHS"))
  fmR<-BGLR(y=y,ETA=ETA,nIter=1000,burnIn=500,thin=10)
  varU=scan("ETA_1_parBayesC.dat")
  varE=scan("varE.dat")
  
  COR.trt<-cor(fmR$yHat,y, use="pairwise.complete.obs")
  if(i==1){CORst<-COR.trt}
  if(i!=1){CORst<-cbind(CORst,COR.trt)}
  
  COR.tst<-cor(fmR$yHat[whichNa],y[whichNa], use="pairwise.complete.obs")
  if(i==1){CORs<-COR.tst}
  if(i!=1){CORs<-cbind(CORs,COR.tst)}
  
  h2<-varU/(varU+varE)
  if(i==1){hs<-mean(h2,na.rm=T)}
  if(i!=1){hs<-cbind(hs,mean(h2,na.rm=T))}
  
  MSE.trn<-mean(na.omit((fmR$yHat[-whichNa]-y[-whichNa])^2))
  if(i==1){MSE.trn_s<-mean(MSE.trn,na.rm=T)}
  if(i!=1){MSE.trn_s<-cbind(MSE.trn_s,mean(MSE.trn,na.rm=T))}
  
  MSE.tst<-mean(na.omit((fmR$yHat[whichNa]-y[whichNa])^2))
  if(i==1){MSE.tst_s<-mean(MSE.tst,na.rm=T)}
  if(i!=1){MSE.tst_s<-cbind(MSE.tst_s,mean(MSE.tst,na.rm=T))}
  
  GEBVs <- fmR$yHat-fmR$mu
  if(i==1){GEBVst_s<-GEBVs}
  if(i!=1){GEBVst_s<-cbind(GEBVst_s,GEBVs)}
  
}
cvh <- hs


```

&nbsp;

```{r}

CORst

CORs

hs

MSE.trn_s

MSE.tst_s

datsetGP <- as.data.frame(cbind(t(CORst),t(CORs),t(hs),t(MSE.trn_s),t(MSE.tst_s)))

rownames(datsetGP) <- c("CV1","CV2","CV3","CV4","CV5")

colnames(datsetGP) <- c("COR.trn",	"COR.tst",	"h2",	"MSE.trn",	"MSE.tst")


write.table(datsetGP,"D:/AGROSAVIA_ 2020 _BUHO/PAPERS_2020/Korea Case/Paper GP/A someter/American Naturalist/Scripts/Genomic Prediction/datsetGP.csv", row.names = FALSE, sep = "\t")
```

&nbsp;

#### Statistics Heritability using RKHS (Cross Validation)

```{r}
# Mean
mean(cvh)
# Standard Desviation
sd(cvh)
# Quantile
quantile(cvh,c(0.025,0.975))
```

&nbsp;

#### Statistics Modeling using RKHS (Cross Validation)

```{r}
#Correlation
mean(t(CORs))
CORs <- as.data.frame(CORs)
names(CORs) <- c("COR Fold 1","COR Fold 2","COR Fold 3","COR Fold 4","COR Fold 5")
CORs %>%
  kbl(caption = "Correlation of Validation set Across the Foldsx") %>%
  kable_classic(full_width = F, html_font = "Cambria")

# Mean Square Error testing
mean(MSE.tst_s)
MSE.tst_s <- as.data.frame(MSE.tst_s)
names(MSE.tst_s) <- c("MSE.tst Fold 1","MSE.tst Fold 2","MSE.tst Fold 3","MSE.tst Fold 4","MSE.tst Fold 5")
MSE.tst_s %>%
  kbl(caption = "Mean Square Error of Validation Set Across the Foldsx") %>%
  kable_classic(full_width = F, html_font = "Cambria")

```

&nbsp;

#### Plotting Heritability

```{r}
# Defining Dataset
datai2 <- as.data.frame(t(cvh))
colnames(datai2) <- c("V1")
datai2$tag <- c("cv")
# Plotting
ggplot(datai2, aes(x=V1, fill=tag)) +
  # geom_histogram(alpha=.2) +
  geom_density(alpha=.7,bw = "ucv",adjust = 1.25) +
  scale_fill_manual(values=c("seagreen","gold")) +
  # scale_fill_brewer(palette="Dark2") + 
  xlim(0, 1) +
  geom_vline(aes(xintercept=mean(mean(h2))), colour="grey28",linetype="dashed", size=1)+
  geom_vline(aes(xintercept=mean(cvh)), colour="grey28",linetype="dashed", size=1) + 
  theme(legend.position = "none")+labs(title ="", y= "Density", x = "Heredability")
```

&nbsp;

#### Breedig Values

```{r}
GEBVst_s <- as.data.frame(GEBVst_s)
colnames(GEBVst_s) <- c("GEBV Fold 1","GEBV Fold 2","GEBV Fold 3","GEBV Fold 4","GEBV Fold 5")
# rownames(GEBVst_s) <- rownames(X)
# Select Fold
fold <- 3 # 2,3,4 or 5
selectedFold <- as.data.frame(GEBVst_s[,fold])
rownames(selectedFold) <- rownames(X)
selectedFold$Plant <- rownames(X)
colnames(selectedFold) <- c("Breedig Value","plant")
Bv <- selectedFold[order(-selectedFold$`Breedig Value`),] 
Bv %>%
  kbl(caption = "") %>%
  kable_classic(full_width = F, html_font = "Cambria")

write.table(Bv,"D:/AGROSAVIA_ 2020 _BUHO/PAPERS_2020/Korea Case/Paper GP/A someter/American Naturalist/Scripts/Genomic Prediction/YLP_Environment2.csv", row.names = FALSE, sep = "\t")

```

&nbsp;

## BRR (Bayesian Ridge Regression) 

&nbsp;

#### Cross-validation using BRR
&nbsp;

```{r}
Environment <- 2 # 1, 2 or 3
y<-YLP[,Environment] # NP, YLP, NS, 
```

```{r }
set.seed(0000)
sets <- sample(c(1,2,3,4,5), size = 83, replace = TRUE)
#Case III - Cross-validation 
for(i in 1:5){
  yNa=y
  whichNa=(sets==i)
  yNa[whichNa]=NA
  
  ETA=list(list(X=X,model="BRR"))
  # Z=scale(X,center=TRUE,scale=TRUE)
  # G=tcrossprod(Z)/ncol(Z)
  # ETA=list(list(K=G,model="BRR"))
  fmR<-BGLR(y=yNa,ETA=ETA,nIter=1000,burnIn=500,thin=10)
  varU=scan("ETA_1_parBayesC.dat")
  varE=scan("varE.dat")
  
  COR.tst<-cor(fmR$yHat[whichNa],y[whichNa], use="pairwise.complete.obs")
  if(i==1){CORs<-COR.tst}
  if(i!=1){CORs<-cbind(CORs,COR.tst)}
  
  h2<-varU/(varU+varE)
  if(i==1){hs<-mean(h2,na.rm=T)}
  if(i!=1){hs<-cbind(hs,mean(h2,na.rm=T))}
  
  MSE.trn<-mean(na.omit((fmR$yHat[-whichNa]-y[-whichNa])^2))
  if(i==1){MSE.trn_s<-mean(MSE.trn,na.rm=T)}
  if(i!=1){MSE.trn_s<-cbind(MSE.trn_s,mean(MSE.trn,na.rm=T))}
  
  MSE.tst<-mean(na.omit((fmR$yHat[whichNa]-y[whichNa])^2))
  if(i==1){MSE.tst_s<-mean(MSE.tst,na.rm=T)}
  if(i!=1){MSE.tst_s<-cbind(MSE.tst_s,mean(MSE.tst,na.rm=T))}
  
}
cvh <- hs
```

&nbsp;

```{r}
CORs

hs

MSE.trn_s

MSE.tst_s

```

&nbsp;

#### Statistics Modeling using BRR (Cross Validation)

```{r}
#Correlation
mean(t(CORs))
CORs <- as.data.frame(CORs)
names(CORs) <- c("COR Fold 1","COR Fold 2","COR Fold 3","COR Fold 4","COR Fold 5")
CORs %>%
  kbl(caption = "Correlation of Validation set Across the Foldsx") %>%
  kable_classic(full_width = F, html_font = "Cambria")

# Mean Square Error testing
mean(MSE.tst_s)
MSE.tst_s <- as.data.frame(MSE.tst_s)
names(MSE.tst_s) <- c("MSE.tst Fold 1","MSE.tst Fold 2","MSE.tst Fold 3","MSE.tst Fold 4","MSE.tst Fold 5")
MSE.tst_s %>%
  kbl(caption = "Mean Square Error of Validation Set Across the Foldsx") %>%
  kable_classic(full_width = F, html_font = "Cambria")

```

&nbsp;

#### Marker effects - Betas 

&nbsp;

#### Selecting the best Moldel using the fold with best MSE and COR

```{r}
yNa=y
whichNa=(sets==3) # 1,2,3,4 or 5
```

&nbsp;

####

```{r include=FALSE}
yNa[whichNa]=NA

ETA=list(list(X=X,model="BL"))
fmR<-BGLR(y=yNa,ETA=ETA,nIter=1000,burnIn=500,thin=10)
varU=scan("ETA_1_parBayesC.dat")
varE=scan("varE.dat")
```

&nbsp;

#### Mean Square Error testing using BRR

```{r}
# Testing
MSE.tst<-mean(na.omit((fmR$yHat[whichNa]-y[whichNa])^2))
MSE.tst
# Training
MSE.trn<-mean(na.omit((fmR$yHat[-whichNa]-y[-whichNa])^2))
MSE.trn
```

&nbsp;

#### Correlation using BRR

```{r}
# Testing
COR.tst<-cor(fmR$yHat[whichNa],y[whichNa], use="pairwise.complete.obs") #at least 0.3!
COR.tst
# Training
COR.trn<-cor(fmR$yHat[-whichNa],y[-whichNa], use="pairwise.complete.obs")
COR.trn
```

&nbsp;

#### Plotting Correlation using BRR

```{r}
#R
par(mfrow=c(1,1))
plot(y,fmR$yHat,xlab="Phenotype",ylab="Pred. Gen. Value" ,cex=1,bty="L",pch=1)
points(x=y[whichNa],y=fmR$yHat[whichNa],col="royalblue1",cex=1,pch=19)
legend("topleft", legend=c("training","testing"),bty="n",pch=c(1,19),col=c("black","royalblue1"))
```

&nbsp;

#### Marker effects using BRR

```{r}
#M - Betas - Only for BRR
beta_hat<-fmR$ETA[[1]]$b
plot(beta_hat,col="royalblue1",cex=.7,pch=1)
```

#### Top 50 Marker effects using BRR

```{r}
#M - Betas - Only for BRR
Betas <- as.data.frame(head(sort(beta_hat,decreasing=T),n=-50))
colnames(Betas) <- "beta"
Betas <- cbind(rownames(Betas),Betas$beta)
colnames(Betas) <- c("Molecuar Variant","beta")

Betas %>%
  kbl(caption = "Top 50 Marker effects using BRR") %>%
  kable_classic(full_width = F, html_font = "Cambria")
```



