---
title: "Genomic Prediction"
output: html_notebook
---

&nbsp;


&nbsp;

#### Uploading R libraries

```{r}
library(BGLR)
library(ggplot2)
library(readr)
library(rcompanion)
library(readxl)
library(kableExtra)
library(GAPIT)
```

&nbsp;

#### Analysis Path

```{r}
setwd("D:/AGROSAVIA_ 2020 _BUHO/PAPERS_2020/Korea Case/Paper GP/A someter/American Naturalist/Scripts/Genomic Prediction")
```

&nbsp;

#### Uploading the Allelic Variant File 

```{r}
library(GAPIT)
# Using GAPIT for convert the hmap format to numeric format
# Set the folder relative to the variable, in this case we use "yield" as an example. Explore the folders: "NP" (number of pods), "NS" (average number of seeds per pod), "BS" (seed weight), and "VB" (vegetative biomass )
# myG <- as.data.frame(read_delim("D:/AGROSAVIA_ 2020 _BUHO/PAPERS_2020/Korea Case/Paper GP/Genotype/Yield/Yield_KKNimp.hmp.txt",
#                                 delim = "\t", escape_double = FALSE,
#                                 col_names = FALSE, trim_ws = TRUE))
# myGAPIT <- GAPIT(G=myG, output.numerical=TRUE,
#                  Geno.View.output=FALSE)

# In this case, we use the top markers generated in previous steps. We use the example of 1,000 markers, but there are between 25 top markers and up to 10,000 top markers available (in the path ./Genotype/"variable", in this case "Yield")
myG <- as.data.frame(read_delim("D:/AGROSAVIA_ 2020 _BUHO/PAPERS_2020/Korea Case/Paper GP/A someter/American Naturalist/Scripts/Genomic Prediction/1000_YLP_Turipana.hmp.txt",
                                delim = "\t", escape_double = FALSE,
                                col_names = FALSE, trim_ws = TRUE))
myGAPIT <- GAPIT(G=myG, output.numerical=TRUE,
                 Geno.View.output=FALSE)

# myG <- as.data.frame(read_delim("D:/AGROSAVIA_ 2020 _BUHO/PAPERS_2020/Korea Case/Paper GP/Genotype/Yield/Yield_GWAS_KKNimp.hmp.txt",
#                                 delim = "\t", escape_double = FALSE,
#                                 col_names = FALSE, trim_ws = TRUE))
# myGAPIT <- GAPIT(G=myG, output.numerical=TRUE,
#                  Geno.View.output=FALSE)

```

```{r}
# Set the folder relative to the variable in the path, in this case we use the folder "yield" as an example. Explore the folders: "NP" (number of pods), "NS" (average number of seeds per pod), "BS" (seed weight), and "VB" (vegetative biomass )
X <- read.delim("D:/AGROSAVIA_ 2020 _BUHO/PAPERS_2020/Korea Case/Paper GP/A someter/American Naturalist/Scripts/Genomic Prediction/GAPIT.Genotype.Numerical.txt", row.names=1)


X[X==2] <- -1
X[X==1] <- 7
X[X==0] <- 1
X[X==7] <- 0

X[1:10,1:10] %>%
  kbl(caption = "Allelic Variant ") %>%
  kable_classic(full_width = F, html_font = "Cambria")
dim(X)
```

&nbsp;

#### Uploading the Genomic Relationship matrix
```{r include=FALSE} 
G <- read_csv("D:/AGROSAVIA_ 2020 _BUHO/PAPERS_2020/Korea Case/Paper GP/A someter/American Naturalist/Scripts/Genomic Prediction/GAPIT.Kin.VanRaden.csv", 
              col_names = FALSE)
G <- as.matrix(G[,-1])
```

```{r}
G[1:5,1:5] %>%
  kbl(caption = "Genomic Relationship matrix") %>%
  kable_classic(full_width = F, html_font = "Cambria")
```

&nbsp;

Upload the phenotypic data related to variable "x", we use "Yield" as an example
```{r}
# Set the folder relative to the variable in the path, in this case we use the folder "yield" as an example. Explore the folders: "NP" (number of pods), "NS" (average number of seeds per pod), "BS" (seed weight), and "VB" (vegetative biomass )
YLP1 <- read_delim("D:/AGROSAVIA_ 2020 _BUHO/PAPERS_2020/Korea Case/Paper GP/A someter/American Naturalist/Scripts/Genomic Prediction/Index_YLP_Kolfaci.txt", 
    delim = "\t", escape_double = FALSE, 
    trim_ws = TRUE)
names <- as.data.frame(rownames(X))
colnames(names) <- c("Taxa")
YLP <- merge(YLP1, names, by="Taxa")
YLP <- YLP[,1:5]
YLP %>%
  kbl(caption = "Phenotypic Table Yield") %>%
  kable_classic(full_width = F, html_font = "Cambria")
```


&nbsp;

#### Cross-validation

&nbsp;

```{r}
# For example, Environment 2 refers to the second column related to the environment 'Motilonia'.
Environment <- 2
# 1, 2 or 3
y<-YLP[,Environment]
```

```{r}
set.seed(0000)# We use several seeds in the study to measure the stochastic outputs
sets <- sample(c(1,2,3,4,5), size = 83, replace = TRUE)
#Case III - Cross-validation 
for(i in 1:5){
  yNa=y
  whichNa=(sets==i)
  yNa[whichNa]=NA
  
  ETA=list(list(X=X,model="BayesC"))
  # Z2 <- as.matrix(X)
  # G2= tcrossprod(Z2)/ncol(Z2)
  # ETA=list(list(K=G2,model="RKHS"))
  fmR<-BGLR(y=y,ETA=ETA,nIter=1000,burnIn=500,thin=10)
  varU=scan("ETA_1_parBayesC.dat")
  varE=scan("varE.dat")
  
  COR.trt<-cor(fmR$yHat,y, use="pairwise.complete.obs")
  if(i==1){CORst<-COR.trt}
  if(i!=1){CORst<-cbind(CORst,COR.trt)}
  
  COR.tst<-cor(fmR$yHat[whichNa],y[whichNa], use="pairwise.complete.obs")
  if(i==1){CORs<-COR.tst}
  if(i!=1){CORs<-cbind(CORs,COR.tst)}
  
  h2<-varU/(varU+varE)
  if(i==1){hs<-mean(h2,na.rm=T)}
  if(i!=1){hs<-cbind(hs,mean(h2,na.rm=T))}
  
  MSE.trn<-mean(na.omit((fmR$yHat[-whichNa]-y[-whichNa])^2))
  if(i==1){MSE.trn_s<-mean(MSE.trn,na.rm=T)}
  if(i!=1){MSE.trn_s<-cbind(MSE.trn_s,mean(MSE.trn,na.rm=T))}
  
  MSE.tst<-mean(na.omit((fmR$yHat[whichNa]-y[whichNa])^2))
  if(i==1){MSE.tst_s<-mean(MSE.tst,na.rm=T)}
  if(i!=1){MSE.tst_s<-cbind(MSE.tst_s,mean(MSE.tst,na.rm=T))}
  
  GEBVs <- fmR$yHat-fmR$mu
  if(i==1){GEBVst_s<-GEBVs}
  if(i!=1){GEBVst_s<-cbind(GEBVst_s,GEBVs)}
  
  if(i==3){ ##Please select de best fold in the raw table from paper
    beta_hat<-fmR$ETA[[1]]$b
# plot(beta_hat,col="royalblue1",cex=.7,pch=1)
#M - Betas - Only for BRR
    Betas <- as.data.frame(sort(beta_hat,decreasing=T))
    colnames(Betas) <- "beta"
    Betas <- cbind(rownames(Betas),Betas$beta)
    colnames(Betas) <- c("Molecuar Variant","beta")
    write.table(Betas,"D:/AGROSAVIA_ 2020 _BUHO/PAPERS_2020/Korea Case/Paper GP/SNP Saturation/Betas.csv", row.names = FALSE, sep = "\t")
  }
  
}
cvh <- hs


```


&nbsp;

```{r}

CORst

CORs

hs

MSE.trn_s

MSE.tst_s

datsetGP <- as.data.frame(cbind(t(CORst),t(CORs),t(hs),t(MSE.trn_s),t(MSE.tst_s)))

rownames(datsetGP) <- c("CV1","CV2","CV3","CV4","CV5")

colnames(datsetGP) <- c("COR.trn",	"COR.tst",	"h2",	"MSE.trn",	"MSE.tst")


# write.table(datsetGP,"D:/AGROSAVIA_ 2020 _BUHO/PAPERS_2020/Korea Case/Paper GP/SNP Saturation/datsetGP.csv", row.names = FALSE, sep = "\t")
```

```{r}
round(mean(datsetGP$COR.tst), digits = 3)
```

